File: python/binarySearch/binarySearch.py
import sys
import threading
from typing import List, Tuple, Dict, Set
from collections import defaultdict, deque

# from heapq import heappush, heappop  # For priority queues


class Solution:
    def searchRecursive(nums, target, left, right):
        if left > right:
            return -1
        mid = (right + left) // 2
        if nums[mid] == target:
            return mid
        if nums[mid] > target:
            return searchRecursive(nums, target, left, mid - 1)
        else:
            return searchRecursive(nums, target, mid + 1, right)

    def search(self, nums: List[int], target: int) -> int:
        left = 0
        right = len(nums) - 1
        while left <= right:
            mid = left + (right - left) // 2
            if nums[mid] == target:
                return mid
            if nums[mid] < target:
                left = mid + 1
            else:
                right = mid - 1
        return -1


def main():
    s = Solution()
    data = [1, 2, 3]
    print(data[len(data) - 1])
    s.search(data, 3123)


if __name__ == "__main__":
    threading.Thread(target=main).start()

File: cpp/binarySearch/binary_search.cpp
#include <bits/stdc++.h>
using namespace std;

#define ll long long
#define pb push_back
#define all(x) x.begin(), x.end()
#define endl '\n'

int searchRecursive(vector<int>& nums, int target, int left, int right) {
    int mid = (right + left) / 2;
    if (left >= right) return -1;
    if (nums[mid] == target) return mid;
    if (nums[mid] > target)
        return searchRecursive(nums, target, left, mid - 1);
    else
        return searchRecursive(nums, target, mid + 1, right);
}
int search(vector<int>& nums, int target) {
    int left = 0;
    int right = nums.size() - 1;

    while (left <= right) {
        int mid = (right + left) / 2;
        if (nums[mid] == target) return mid;
        if (nums[mid] > target)
            right = mid - 1;
        else
            left = mid + 1;
    }

    return -1;
}

int main() {

    ios::sync_with_stdio(0);
    cin.tie(0);

    return 0;
}
